### PR TITLE
Rework testutils.py

### PR DESCRIPTION
**Pull request by fruffy**
This PR refactors the `testutils` and `ebpfenv` class:
- Make `shell=True` optional so that timeouts can work properly. If the subprocess is executed in a shell, `timeout` can not kill the process properly.
- Simplify the exec_process command and directly log the subprocess outputs to the active log.
- Add more typing information to the testutils and ebpfenv file. The goal is to incrementally add typing information to all the Python code.

Most of the changes to the code are a consequence of these contributions.

### CODE REVIEW

**Comment by fruffy** [comment]
@mbudiu-vmw If you have some time, could you give this a review/approval? It mostly cleans up the Python code and testutils library.

File: backends/bmv2/run-bmv2-ptf-test.py [hunk 1]
```diff
@@ -59,74 +59,94 @@
...
+def run_ptf(options: Options, bridge: Bridge, grpc_port: int, json_name: Path,
+            info_name: Path) -> int:
+    """Run the PTF test."""
+    testutils.log.info("---------------------- Run PTF test ----------------------")
     # Add the file location to the python path.
     pypath = FILE_DIR
     # Show list of the tests
-    testLostCmd = f"ptf --pypath {pypath} --test-dir {testdir} --list"
-    returncode = bridge.ns_exec(testLostCmd)
+    test_list_cmd = f"ptf --pypath {pypath} --test-dir {options.testdir} --list"
```
Comment (mbudiu-vmw) [inline backends/bmv2/run-bmv2-ptf-test.py:141]: lost is not the same as list. which one is right?

File: backends/ebpf/targets/ebpfenv.py [hunk 2]
```diff
@@ -29,66 +29,77 @@
...
 
-    def __init__(self, namespace):
-        self.ns_name = namespace # identifier of the namespace
-        self.br_name = "core"    # name of the central bridge
-        self.br_ports = []       # list of the veth pair bridge ports
-        self.edge_ports = []     # list of the veth pair edge ports
-
-    def ns_init(self):
+    def __init__(self, namespace: str):
+        # Identifier of the namespace.
```
Comment (mbudiu-vmw) [inline backends/ebpf/targets/ebpfenv.py:33]: we have too many bugs to fix to spend the time to reformat the way comments look.
and to review that code. that's a lot of wasted energy.

File: backends/ebpf/targets/ebpfenv.py [hunk 3]
```diff
@@ -101,20 +112,23 @@ def _configure_bridge(self, br_name):
...
-    def create_bridge(self):
+    def create_bridge(self) -> int:
         """Create the central bridge of the environment and configure it."""
         result = self.ns_exec(f"ip link add {self.br_name} type bridge")
         if result != testutils.SUCCESS:
             return result
         return self._configure_bridge(self.br_name)
 
-    def _configure_bridge_port(self, port_name):
+    def _configure_bridge_port(self, port_name: str):
```
Comment (mbudiu-vmw) [inline backends/ebpf/targets/ebpfenv.py:122]: no return type here?

File: backends/ebpf/targets/kernel_target.py [hunk 4]
```diff
@@ -54,8 +54,10 @@ def compile_dataplane(self):
             args += f" INCLUDES+=-include{self.options.extern} "
             # need to include the temporary dir because of the tmp import
             args += f" INCLUDES+=-I{self.tmpdir} "
-        errmsg = "Failed to compile the eBPF byte code:"
-        return testutils.exec_process(args, errmsg).returncode
+        result = testutils.exec_process(args)
+        if result.returncode != testutils.SUCCESS:
```
Comment (mbudiu-vmw) [inline backends/ebpf/targets/kernel_target.py:58]: this repeated code suggests that you want a function that takes a message and a command and logs the message if the command failed

File: backends/p4tools/modules/testgen/targets/bmv2/backend/ptf/ptf.cpp [hunk 5]
```diff
@@ -260,11 +248,18 @@ class AbstractTest(bt.P4RuntimeTest):
         pass
 
     def runTestImpl(self):
-        self.setupCtrlPlane()
-        logger.info("Sending Packet ...")
-        self.sendPacket()
-        logger.info("Verifying Packet ...")
-        self.verifyPackets()
+        # TODO: This is a workaround for broken unit tests on Ubuntu 20.04. It can happen that,
+        # when the last test fails, the failure output of that test is swallowed by the unit test
```
Comment (mbudiu-vmw) [inline backends/p4tools/modules/testgen/targets/bmv2/backend/ptf/ptf.cpp:252]: what is the unit test framework?
where is the bug?

File: tools/testutils.py [hunk 6]
```diff
@@ -57,174 +92,210 @@ def hex_to_byte(hex_str):
...
-        sys.exit(1)
-    input_path = Path(input_path)
+    if not input_path_str:
+        log.error("input_path is None")
+        return None
+    input_path = Path(input_path_str)
     if not input_path.exists():
-        log.error(sys.stderr, f"{input_path} does not exist")
-        sys.exit(1)
+        log.error("%s does not exist", input_path)
```
Comment (mbudiu-vmw) [inline tools/testutils.py:266]: you can't make up your mind on using f strings of %?

**Review by mbudiu-vmw — APPROVED** [review]

File: tools/testutils.py (hunk 6 above)
Comment (fruffy) [reply tools/testutils.py:266]: Yes, actually. It is somewhat of a contentious topic https://stackoverflow.com/questions/34619790/pylint-message-logging-format-interpolation
I think using the lazy logging interpolation with `,` is best practice, even if it uses the old formatting style. For all the other formatting, f-strings should be used.

File: backends/p4tools/modules/testgen/targets/bmv2/backend/ptf/ptf.cpp (hunk 5 above)
Comment (fruffy) [reply backends/p4tools/modules/testgen/targets/bmv2/backend/ptf/ptf.cpp:252]: I will remove this. The bug is in this PR https://github.com/p4lang/p4c/pull/3875 but this approach does not solve it. I can not reproduce these failures on Ubuntu 20.04 or 22.04. It looks like some CI problem. The unit test framework is Python `unittest`, which PTF relies on.

File: backends/ebpf/targets/kernel_target.py (hunk 4 above)
Comment (fruffy) [reply backends/ebpf/targets/kernel_target.py:58]: I weighed off both of these options and I felt that keeping the `exec_process` class simple is more beneficial. Most of these error messages are redundant.

**Pull request merged**